src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -1832,6 +1832,87 @@ public Object clone() throws CloneNotSupportedException {
+     *
+     *  Note that some fields, i.e. the cigar elements, alignment blocks, and
+     *  indexing bin, are not explicitly populated in the copy since they are lazily
+     *  generated on demand.
+     *
+     *  Also note that this fails:
+     *
+     *     original.deepCopy().equals(original)
+     *
+     *  due to the fact that SAMBinaryTagAndValue.equals winds up calling object.equals on the
+     *  value field, which uses reference equality.
+     *
+     */
+    public SAMRecord deepCopy() {
+        final SAMRecord newSAM = new SAMRecord(getHeader());
+
+        newSAM.setReadName(getReadName());
+        newSAM.setReadBases(Arrays.copyOf(getReadBases(), getReadLength()));
+        final byte baseQualities[] = getBaseQualities();
+        newSAM.setBaseQualities(Arrays.copyOf(baseQualities, baseQualities.length));
+        newSAM.setReferenceName(getReferenceName());
+        newSAM.setAlignmentStart(getAlignmentStart()); // sets mAlignmentEnd
+        newSAM.setMappingQuality(getMappingQuality());
+        newSAM.setCigarString(getCigarString()); // clears Cigar element and alignmentBlocks
+        newSAM.setFileSource(null);
+
+        newSAM.setFlags(getFlags());
+        newSAM.setMateReferenceName(getMateReferenceName());
+        newSAM.setMateAlignmentStart(getMateAlignmentStart());
+        newSAM.setInferredInsertSize(getInferredInsertSize());
+        newSAM.setReferenceIndex(getReferenceIndex());
+        newSAM.setMateReferenceName(getMateReferenceName());
+        newSAM.setMateReferenceIndex(getMateReferenceIndex());
+        newSAM.setValidationStringency(getValidationStringency());
+
+        SAMBinaryTagAndValue attributes = getBinaryAttributes();
+        while (null != attributes) {
+                final Object value = attributes.value;
+                final String tag = SAMTagUtil.getSingleton().makeStringTag(attributes.tag);
+                if (value instanceof byte[]) {
+                    if (isUnsignedArrayAttribute(tag)) {
+                        newSAM.setUnsignedArrayAttribute(tag, ((byte[]) value).clone());
+                    }
+                    else {
+                        newSAM.setAttribute(tag, ((byte[]) value).clone());
+                    }
+                }
+                else if (value instanceof short[]) {
+                    if (isUnsignedArrayAttribute(tag)) {
+                        newSAM.setUnsignedArrayAttribute(tag, ((short[]) value).clone());
+                    }
+                    else {
+                        newSAM.setAttribute(tag, ((short[]) value).clone());
+                    }
+                }
+                else if (value instanceof int[]) {
+                    if (isUnsignedArrayAttribute(tag)) {
+                        newSAM.setUnsignedArrayAttribute(tag, ((int[]) value).clone());
+                    }
+                    else {
+                        newSAM.setAttribute(tag, ((int[]) value).clone());
+                    }
```

Comment:
I think all this branching based on whether the attribute is an unsigned array attribute is unnecessary -- all you get by going through `setUnsignedArrayAttribute()` is a check that the object is an array (which we already know) and that its length is non-zero (which we can assume, since we should assume that the existing `SAMRecord` is valid). I think you can just call directly into `setAttribute(final short tag, final Object value, final boolean isUnsignedArray)`, passing in `isUnsignedArrayAttribute(tag)` for the third parameter.